src/main.rs

```diff
@@ -164,8 +169,54 @@ async fn monitor(
             for (handle, _name) in abort_handles {
                 drop(handle);
             }
+            initialize_proxy(router, None, epsilon).await;
+            return;
+        }
+
+        // Check if the proxy state is down, and if so, reinitialize the proxy.
+        if is_proxy_down() {
+            error!("Proxy state is DOWN. Reinitializing proxy...");
+            drop(abort_handles); // Drop all abort handles
+            initialize_proxy(router, None, epsilon).await;
             return;
         }
+
         tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
     }
 }
+
+// To keep track of proxy global state
+#[derive(Debug, Clone, Copy)]
+pub enum ProxyState {
+    Pool(PoolState),
+}
+
+// To keep track of the status of the Pool
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum PoolState {
+    Up,
+    Down,
+}
+
+// Checks the states of Pool, and subsequently (Translator,JD, etc) and updates the ProxyState to show the state.
+fn _update_proxy_state(pool: PoolState) {
+    if PROXY_STATE
+        .safe_lock(|proxy_state| {
+            // If any of the states is "Down", we change the ProxyState to reflect that.
+            *proxy_state = match pool {
+                PoolState::Down => ProxyState::Pool(PoolState::Down),
+                _ => ProxyState::Pool(PoolState::Up),
+            };
+        })
+        .is_err()
+    {
+        error!("Error updating proxy state");
+    }
+}
+
+// Check if the proxy state is down
+fn is_proxy_down() -> bool {
```

Comment:
I thought` is_proxy_down()` would be a better name because the function will later check for other states as well. for example `ProxyState::JDC(JDCState::Down)` or `ProxyState::Translator(TranslatorState::Down)`

Reply:
what do you think? @Fi3